### PR TITLE
V1-162 Feature/course-editor

### PR DESCRIPTION
**Pull request by PechurinaLena**
Implemented CourseEditor with 4 editor tabs: Definition step, Skills step, Lessons step, Test step and transition between them with button `Previous`, `Next`
To open editor, you should login as admin, open `Courses list` page -> click `Details` button -> click `Edit` button

! Form is not validated and not submitted at this branch

### CODE REVIEW

File: frontend/src/pages/CourseEditor/TestStep/TestItem/styled.ts [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+  borderBottom: '1px solid #C6C6C9',
+  '&:after, &:before': {
+    display: 'none',
+  },
+});
+
+export const ButtonBox = styled(Box)({
+  width: '100%',
+  display: 'flex',
+  justifyContent: 'end',
```
Comment (irina-sinitsyna) [inline frontend/src/pages/CourseEditor/TestStep/TestItem/styled.ts:75]: order

File: frontend/src/pages/CourseEditor/types.ts [hunk 2]
```diff
@@ -35,3 +45,38 @@ export interface ISkillsStepProps extends IStepProps {
     [key: string]: { group: string; image: string; maxScore: number; name: string; _id: string };
   };
 }
+
+export interface ILessonsStepContainerProps {
+  formik: IFormik;
+  courseData?: ICourse;
+}
```
Comment (irina-sinitsyna) [inline frontend/src/pages/CourseEditor/types.ts:52]: it's the same as IStepProps

File: frontend/src/pages/CourseEditor/types.ts [hunk 3]
```diff
@@ -35,3 +45,38 @@ export interface ISkillsStepProps extends IStepProps {
...
+  selectOption: string;
+  handleChangeOption?: (event: BaseSyntheticEvent) => void;
+}
+
+export interface ILessonItemProps {
+  formik: IFormik;
+  selectOption: string;
+  handleChangeOption?: (event: BaseSyntheticEvent) => void;
+  material?: IMaterial;
+  id: number;
```
Comment (irina-sinitsyna) [inline frontend/src/pages/CourseEditor/types.ts:75]: order

File: frontend/src/pages/CourseEditor/LessonsStep/LessonItem/styled.ts [hunk 4]
```diff
@@ -0,0 +1,84 @@
+import { styled, Box, Input } from '@mui/material';
+
+export const LessonItemWrapper = styled(Box)({
+  margin: '32px 0 50px 0',
+});
+
```
Comment (irina-sinitsyna) [inline frontend/src/pages/CourseEditor/LessonsStep/LessonItem/styled.ts:6]: wrong UI
image
image

Comment (irina-sinitsyna) [reply frontend/src/pages/CourseEditor/LessonsStep/LessonItem/styled.ts:6]: also in case of removing 2nd lesson, the # of next lesson should be 2 not 3

Comment (irina-sinitsyna) [reply frontend/src/pages/CourseEditor/LessonsStep/LessonItem/styled.ts:6]: add more lessons button is not working, blank page appears

Comment (irina-sinitsyna) [reply frontend/src/pages/CourseEditor/LessonsStep/LessonItem/styled.ts:6]: why all the fields have different font styles and size?

Comment (irina-sinitsyna) [reply frontend/src/pages/CourseEditor/LessonsStep/LessonItem/styled.ts:6]: not all the values changes in formik values, please fix it and provide proper structure

**Review by irina-sinitsyna — CHANGES_REQUESTED** [review]

File: frontend/src/pages/CourseEditor/LessonsStep/LessonItem/LessonItem.tsx [hunk 5]
```diff
@@ -5,84 +5,83 @@ import { ILessonItemProps } from 'pages/CourseEditor/types';
...
-              </SelectItem>
-            </InputSelect>
+              </option>
+            </FieldSelect>
             <InputBox>
               <InputLabel>
                 {selectOption === (ContentElementType.video || ContentElementType.presentation)
-                  ? 'Video URL'
-                  : 'Text URL'}
+                  ? `${Titles.videUrlTitle}`
```
Comment (irina-sinitsyna) [inline frontend/src/pages/CourseEditor/LessonsStep/LessonItem/LessonItem.tsx:44]: typo

File: frontend/src/pages/CourseEditor/LessonsStep/LessonsStep.tsx [hunk 6]
```diff
@@ -20,19 +20,19 @@ const LessonsStep: FC<ILessonsStepProps> = ({
...
-                  id={index}
+                  index={index}
                   material={material}
                   formik={formik}
                   selectOption={selectOption}
                   handleChangeOption={handleChangeOption}
                 />
                 {isLastElem(formik.values.materials, index) ? (
-                  <LessonButton variant="mediumOutlined" onClick={() => push({})}>
+                  <LessonButton variant="mediumOutlined" onClick={() => push(material)}>
```
Comment (irina-sinitsyna) [inline frontend/src/pages/CourseEditor/LessonsStep/LessonsStep.tsx:35]: why not empty object?

File: frontend/src/pages/CourseEditor/TestStep/TestItem/TestItem.tsx [hunk 7]
```diff
@@ -14,53 +14,46 @@ import {
...
   InputAnswer,
-  ButtonBox,
-  QuestionActionButton,
   InputText,
   TestBasicField,
-  InputSelect,
+  FieldSelect,
 } from './styled';
 
 const TestItem: FC<ITestItemProps> = ({ formik, handleChangeAnswer, isCurrentAnswer }) => (
```
Comment (irina-sinitsyna) [inline frontend/src/pages/CourseEditor/TestStep/TestItem/TestItem.tsx:22]: Please check this one: https://formik.org/docs/examples/radio-group
I suppose, you don't need to use `isCurrentAnswer` useState here

File: frontend/src/pages/CourseEditor/LessonsStep/LessonItem/LessonItem.tsx [hunk 8]
```diff
@@ -1,87 +1,79 @@
...
-            )}
-            <InputLabel>{Titles.exerciseTitle}</InputLabel>
-            <InputText
-              id={`materials[${index}].exercise.title`}
-              name={`materials[${index}].exercise.title`}
-              placeholder={material.exercise?.title}
-              value={material.exercise?.title}
+const LessonItem: FC<ILessonItemProps> = ({ formik, material, index }) => {
+  return (
+    <LessonItemWrapper>
```
Comment (irina-sinitsyna) [inline frontend/src/pages/CourseEditor/LessonsStep/LessonItem/LessonItem.tsx]: no need to use return here

File: frontend/src/pages/CourseEditor/LessonsStep/LessonItem/LessonItem.tsx [hunk 9]
```diff
@@ -1,87 +1,79 @@
...
+      <ItemTitle>
+        {Titles.lessonCount} {index + 1}
```
Comment (irina-sinitsyna) [inline frontend/src/pages/CourseEditor/LessonsStep/LessonItem/LessonItem.tsx]: magic number (please move to constants)

File: frontend/src/pages/CourseEditor/LessonsStep/LessonItem/LessonItem.tsx [hunk 10]
```diff
@@ -1,87 +1,79 @@
...
+      </ItemTitle>
+      <LessonInnerBox>
+        <Field
+          select
+          value={material.type}
+          id={`materials[${index}].type`}
+          name={`materials[${index}].type`}
+          onChange={formik.handleChange}
+        >
+          {LESSONS_TYPE.map((option) => (
```
Comment (irina-sinitsyna) [inline frontend/src/pages/CourseEditor/LessonsStep/LessonItem/LessonItem.tsx]: let's replace `option` with something more descriptive (type / typeOption ? )

File: frontend/src/pages/CourseEditor/LessonsStep/LessonsStep.tsx [hunk 11]
```diff
@@ -10,13 +10,7 @@ import { FormWrapper, SectionName } from 'pages/CourseEditor/styled';
...
-  formik,
-  lessonsContent,
-  courseData,
-  selectOption,
-  handleChangeOption,
-}) => (
+const LessonsStep: FC<ILessonsStepProps> = ({ formik, courseData }) => (
   <FormWrapper>
     <SectionName>{Titles.lessonStepTitle}</SectionName>
     {courseData &&
```
Comment (irina-sinitsyna) [inline frontend/src/pages/CourseEditor/LessonsStep/LessonsStep.tsx:16]: please add a loader here as in skills and definition steps

File: frontend/src/pages/CourseEditor/LessonsStep/LessonsStepContainer.tsx [hunk 12]
```diff
@@ -1,36 +1,11 @@
...
-      lessonsContent={lessonsContent}
-      courseData={courseData}
-      handleChangeOption={handleChangeOption}
-      selectOption={selectOption}
-    />
-  );
-};
+const LessonsStepContainer: FC<IStepProps> = ({ formik, courseData }) => (
+  <LessonsStep formik={formik} courseData={courseData} />
+);
```
Comment (irina-sinitsyna) [inline frontend/src/pages/CourseEditor/LessonsStep/LessonsStepContainer.tsx:9]: so we don't need Container if we don't have any logic here. please remove it

Comment (irina-sinitsyna) [reply frontend/src/pages/CourseEditor/LessonsStep/LessonsStepContainer.tsx:9]: and don't forget to change an export in index.ts

File: frontend/src/pages/CourseEditor/types.ts [hunk 13]
```diff
@@ -54,26 +61,19 @@ export interface IStepProps {
...
 
 export interface ILessonItemProps {
   formik: IFormik;
-  material: IMaterial;
-  lessonsContent: {
-    [index: number]: { type: string; material: string };
+  material: {
+    type: string;
+    material: string;
+    exercise?: { eN: number; title: string; task: string; code: string };
```
Comment (irina-sinitsyna) [inline frontend/src/pages/CourseEditor/types.ts:79]: should be improved as you have 2 similar interfaces

File: frontend/src/pages/CourseEditor/CourseEditor.tsx [hunk 14]
```diff
@@ -15,52 +14,25 @@ import TestStep from './TestStep';
...
-              formik={formik}
-              isCourseDataLoading={isCourseDataLoading}
-            />
-          </EditorTabs>
-        </InnerWrapper>
-      )}
-    </PageTitle>
-  );
-};
+const CourseEditor: React.FC<ICourseEditorProps> = ({ courseData, ...props }) => (
```
Comment (irina-sinitsyna) [inline frontend/src/pages/CourseEditor/CourseEditor.tsx:17]: pleace remove React here

File: frontend/src/pages/CourseEditor/SkillsStep/SkillsStep.tsx [hunk 15]
```diff
@@ -0,0 +1,95 @@
...
+    <Loader type="content" />
+  ) : (
+    <FormWrapper>
+      <SectionName>{EditorTitles.skillStepTitile}</SectionName>
+      <SkillsTitleWrapper>
+        <SkillsText>{EditorTitles.skillDescription}</SkillsText>
+      </SkillsTitleWrapper>
+      {courseData &&
+        courseData.allSkills.length &&
+        formik.values.technologies.map((technology, index) => (
```
Comment (ostapdans99) [inline frontend/src/pages/CourseEditor/SkillsStep/SkillsStep.tsx:40]: map should be inside FieldArray

**Review by irina-sinitsyna — APPROVED** [review]

**Review by dapodev — APPROVED** [review]

**Pull request merged**